util/scheduled_executor.go

```diff
@@ -5,26 +5,75 @@ import (
 	"time"
 )
 
+type CustomTicker interface {
+	Stop()
+	Start()
+	C() <-chan time.Time
+}
+
+type defaultTicker struct {
+	*time.Ticker
+	period time.Duration
+}
+
+func (d *defaultTicker) C() <-chan time.Time {
+	return d.Ticker.C
+}
+
+func (d *defaultTicker) Start() {
+	d.Ticker = time.NewTicker(d.period)
+}
+
 type ScheduledExecutor interface {
 	Start()
 	Stop()
 }
 
 type scheduledExecutor struct {
-	period    time.Duration
-	cb        func()
-	isRunning AtomicBool
-	ticker    *time.Ticker
-	done      chan bool
-	wg        sync.WaitGroup
+	cb           func()
+	isRunning    AtomicBool
+	ticker       CustomTicker
+	done         chan bool
+	wg           sync.WaitGroup
+	initialDelay time.Duration
 }
 
+type Option func(*scheduledExecutor)
+
 func NewScheduledExecutor(period time.Duration, cb func()) ScheduledExecutor {
 	return &scheduledExecutor{
-		period:    period,
-		cb:        cb,
-		isRunning: NewAtomicBool(false),
-		wg:        sync.WaitGroup{},
+		cb:           cb,
+		isRunning:    NewAtomicBool(false),
+		wg:           sync.WaitGroup{},
+		initialDelay: period,
+		ticker:       CustomTicker(&defaultTicker{period: period}),
+	}
+}
+
+func NewScheduledExecutorWithOpts(period time.Duration, cb func(), options ...Option) ScheduledExecutor {
+	s := &scheduledExecutor{
+		cb:           cb,
+		isRunning:    NewAtomicBool(false),
+		wg:           sync.WaitGroup{},
+		initialDelay: period,
+		ticker:       CustomTicker(&defaultTicker{period: period}),
+	}
+
+	for _, opt := range options {
+		opt(s)
+	}
+	return s
+}
+
+func WithInitialDelay(initialDelay time.Duration) Option {
+	return func(s *scheduledExecutor) {
+		s.initialDelay = initialDelay
+	}
+}
```

Comment:
Now that we can specify a custom ticker... do we need to directly support this? Couldn't I just make a ticker that has an initial delay?

Reply:
that's a good point! i just figure if it's going to be used by multiple services, it might be easier to already have it specified here

Reply:
Initial delay complicates the implementation of this scheduled executor a bit. It would be nice if we could separate out that functionality into a custom ticker (which could be provided in this package). 

That way all ScheduledExecutor needs to know is that it will receive a ticker, and that it should run the executor when that thing ticks.